controller/konnect/reconciler_generic.go

```diff
@@ -0,0 +1,263 @@
+func (r *KonnectEntityReconciler[T, TEnt]) SetupWithManager(mgr ctrl.Manager) error {
+	var (
+		e   T
+		ent = TEnt(&e)
+		b   = ctrl.NewControllerManagedBy(mgr).
+			For(ent).
+			Named(entityTypeName[T]()).
+			WithOptions(controller.Options{
+				MaxConcurrentReconciles: MaxConcurrentReconciles,
+			})
+	)
+
+	for _, dep := range ReconciliationWatchOptionsForEntity(r.Client, ent) {
+		b = dep(b)
+	}
+	return b.Complete(r)
+}
+
+// Reconcile reconciles the given Konnect entity.
+func (r *KonnectEntityReconciler[T, TEnt]) Reconcile(
+	ctx context.Context, req ctrl.Request,
+) (ctrl.Result, error) {
+	var (
+		entityTypeName = entityTypeName[T]()
+		logger         = log.GetLogger(ctx, entityTypeName, r.DevelopmentMode)
+	)
+
+	var (
+		e   T
+		ent = TEnt(&e)
+	)
+	if err := r.Client.Get(ctx, req.NamespacedName, ent); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	log.Debug(logger, "reconciling", ent)
+	var (
+		apiAuth    configurationv1alpha1.KonnectAPIAuthConfiguration
+		apiAuthRef = types.NamespacedName{
+			Name:      ent.GetKonnectAPIAuthConfigurationRef().Name,
+			Namespace: ent.GetNamespace(),
+			// TODO(pmalek): enable if cross namespace refs are allowed
+			// Namespace: ent.GetKonnectAPIAuthConfigurationRef().Namespace,
+		}
+	)
+	if err := r.Client.Get(ctx, apiAuthRef, &apiAuth); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get KonnectAPIAuthConfiguration: %w", err)
+	}
+
+	// Check if the referenced APIAuthConfiguration is valid.
+	if cond, present := k8sutils.GetCondition(KonnectAPIAuthConfigurationValidConditionType, &apiAuth.Status); present && cond.Status != metav1.ConditionTrue {
+		// If the referenced APIAuthConfiguration is not valid, set the
+		// "APIAuthRefValid" condition to false and return.
+		k8sutils.SetCondition(
+			k8sutils.NewConditionWithGeneration(
+				KonnectEntityAPIAuthConfigurationRefValidConditionType,
+				metav1.ConditionFalse,
+				KonnectEntityAPIAuthConfigurationRefReasonInvalid,
+				fmt.Sprintf("referenced KonnectAPIAuthConfiguration %s is invalid", apiAuthRef),
+				ent.GetGeneration(),
+			),
+			ent,
+		)
+
+		return ctrl.Result{}, nil
+	}
+
+	// If the referenced APIAuthConfiguration is valid, set the "APIAuthRefValid"
+	// condition to true.
+	k8sutils.SetCondition(
+		k8sutils.NewConditionWithGeneration(
+			KonnectEntityAPIAuthConfigurationRefValidConditionType,
+			metav1.ConditionTrue,
+			KonnectEntityAPIAuthConfigurationRefReasonValid,
+			fmt.Sprintf("referenced KonnectAPIAuthConfiguration %s is valid", apiAuthRef),
+			ent.GetGeneration(),
+		),
+		ent,
+	)
+	if err := r.Client.Status().Update(ctx, ent); err != nil {
+		if k8serrors.IsConflict(err) {
+			return ctrl.Result{Requeue: true}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to update status with %s condition: %w", KonnectEntityAPIAuthConfigurationRefValidConditionType, err)
+	}
```

Comment:
We should return in case the resource has been updated. The update will trigger another reconciliation loop.

Reply:
This is now handled slightly better with `updateStatusWithCondition` PTAL

https://github.com/Kong/gateway-operator/pull/446/files#diff-5b6aadc0fe9b4d0ba8cfd31756eef41ffdcd65687cd077d3b8d427294dcc81b5R300-R331